packages/allure-playwright/src/index.ts

```diff
@@ -201,12 +207,48 @@ class AllureReporter implements Reporter {
         fileName = runtime.writeAttachmentFromPath(attachment.path!, attachment.contentType);
       }
 
-      if (attachment.name.endsWith("-expected.png")) {
-        allureTest.addAttachment("expected", attachment.contentType, fileName);
-      } else if (attachment.name.endsWith("-actual.png")) {
-        allureTest.addAttachment("actual", attachment.contentType, fileName);
-      } else if (attachment.name.endsWith("-diff.png")) {
-        allureTest.addAttachment("diff", attachment.contentType, fileName);
+      if (
+        attachment.name.endsWith("-expected.png") ||
+        attachment.name.endsWith("-actual.png") ||
+        attachment.name.endsWith("-diff.png")
+      ) {
+        const diffEndRegexp = /-((expected)|(diff)|(actual))\.png/;
+
+        const pathWithoutEnd = attachment.path!.replace(diffEndRegexp, "");
+        if (this.processedDiffs.includes(pathWithoutEnd)) {
+          continue;
+        }
+
+        let actualBase64: string | undefined,
+          expectedBase64: string | undefined,
+          diffBase64: string | undefined;
+
+        try {
+          actualBase64 = fs.readFileSync(`${pathWithoutEnd}-actual.png`).toString("base64");
+        } catch (e) {}
+
+        try {
+          expectedBase64 = fs.readFileSync(`${pathWithoutEnd}-expeceeeted.png`).toString("base64");
+        } catch (e) {}
+
+        try {
+          diffBase64 = fs.readFileSync(`${pathWithoutEnd}-diff.png`).toString("base64");
+        } catch (e) {}
+
+        const diffName = attachment.name.replace(diffEndRegexp, "");
+        const res = this.allureRuntime?.writeAttachment(
+          JSON.stringify({
+            expected: expectedBase64 && `data:image;base64,${expectedBase64}`,
```

Comment:
It would be nice, to move the logic to factory-function, what will generate expected string.
By the way, you can read the file by path in the function: `readImageAsBase64(path) -> string | undefined`

Reply:
And IMO better to use promisified versions of `fs` utilities